cJSON.c

```diff
@@ -452,20 +431,22 @@ static cJSON_bool print_number(const cJSON * const item, printbuffer * const out
     {
         length = sprintf((char*)number_buffer, "null");
     }
-    else if ((fabs(floor(d) - d) <= DBL_EPSILON) && (fabs(d) < 1.0e60))
```

Comment:
Since -0 is allowed, there is no need to change it to 0.

As for printing integers without engineering notation as long as they can be printed with full precision, the code as in the pull request mostly handles this.  2^53 is 9,007,199,254,740,992, a little less than 10^17 (17 decimal places). The %1.17g format already prints 17 decimal places.

9007199254740991, 9007199254740992, and 9007199254740994 all print as 16-digit integers as desired.  (9007199254740993 is not representable exactly as a double.)

The problem is with the test-print to 15 digits (%1.15g).  It causes integers ending in 0 that require more than 15 decimal digits to print in exponential notation.

100,000,000,000,000 prints as desired, with 15 decimal digits:  100000000000000
(I manually added comma separators to help myself count.)
1,000,000,000,000,000 prints as 1e+15, but we want 1000000000000000
10,000,000,000,000,000 prints as 1e+16, but we want 10000000000000000
100,000,000,000,000,000 prints as 1e+17, but that is OK according to your proposed rule because it is greater than 2^53.

The solution is to qualify the test-print. I will have to think more carefully about the value of LIMIT_FOR_TEST_PRINTING - is it 1e+15? I'll have more time to think tomorrow night.

        if (fabs(d) < LIMIT_FOR_TEST_PRINTING)
        {
            /* Try 15 decimal places of precision to avoid nonsignificant nonzero digits */
            length = sprintf((char*)number_buffer, "%1.15g", d);

            /* Check whether the original double can be recovered */
            if ((sscanf((char*)number_buffer, "%lg", &test) != 1) || ((double)test != d))
            {
                /* If not, print with 17 decimal places of precision */
                length = sprintf((char*)number_buffer, "%1.17g", d);
            }
        }
        else
        {
            /* Print with 17 decimal places of precision */
            length = sprintf((char*)number_buffer, "%1.17g", d);
        }